### PR TITLE
QuadTree

### PR DESCRIPTION
**Pull request by anikabagga**
New Files:
QuadTree.java
QuadTreeTest.java
BoundingBox.java

### CODE REVIEW

File: capstone/src/test/java/com/google/sps/objects/GroupLocationTest.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  @Test
+  public void findMidPointTest() {
+    Coordinate coordReturned = groupLocation.findMidPoint(groupCoordinates);
+
+    assertTrue(coordReturned.getLat() == MID_LAT);
+    assertTrue(coordReturned.getLng() == MID_LNG);
+  }
+}
```
Comment (himaniyadav) [inline capstone/src/test/java/com/google/sps/objects/GroupLocationTest.java]: What's the functional difference between findGroupMidPoint and findMidpoint?

Comment (himaniyadav) [reply capstone/src/test/java/com/google/sps/objects/GroupLocationTest.java]: Also, it might be good to add a test with a user that doesn't have their address/coordinates set, and check if it still returns the same group midpoint. unless that's not implemented yet

Comment (lxcyqx) [reply capstone/src/test/java/com/google/sps/objects/GroupLocationTest.java]: I believe findMidpoint is a helper method for findGroupMidpoint, in which case findMidpoint can just be made private and then there wouldn't need to be a test for findMidpoint

Comment (anikabagga) [reply capstone/src/test/java/com/google/sps/objects/GroupLocationTest.java]: Yep findMidPoint is a helper method. Makes sense to remove the extra test for findMidPoint. 
Added a test that checks if the same group midpoint is returned with missing user coordinates.

File: capstone/src/main/webapp/index.html [hunk 2]
```diff
@@ -8,6 +8,9 @@
     <link rel="stylesheet" href="styles/login.css">
     <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto">
     <script src="scripts/auth.js"></script>
+    <script
+      src="https://maps.googleapis.com/maps/api/js"
+    ></script>
```
Comment (nicorod-google) [inline capstone/src/main/webapp/index.html]: I don't see any JS code that takes this script tag and appends the api key from the json to it. Is this working as intended? If you have offline code where you are doing this, you should include that so that others can use it and the only thing they would need to run it is the json file.

File: capstone/src/main/java/com/google/sps/Objects/BoundingBox.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    if (circleDistanceY > distance + boundaryHeight/2) {
+      return false;
+    }
+    if (circleDistanceX <= boundaryWidth/2) {
+      return true;
+    }
+    if (circleDistanceY <=boundaryHeight/2) {
+      return true;
+    }
```
Comment (nicorod-google) [inline capstone/src/main/java/com/google/sps/Objects/BoundingBox.java]: nit: add a space before/after the division

File: capstone/src/main/java/com/google/sps/Objects/BoundingBox.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    if (circleDistanceX > distance + boundaryWidth/2) {
+      return false;
+    }
+    if (circleDistanceY > distance + boundaryHeight/2) {
+      return false;
+    }
+    if (circleDistanceX <= boundaryWidth/2) {
+      return true;
+    }
+    if (circleDistanceY <=boundaryHeight/2) {
```
Comment (nicorod-google) [inline capstone/src/main/java/com/google/sps/Objects/BoundingBox.java]: nit: add a space after <=

File: capstone/src/main/java/com/google/sps/Objects/BoundingBox.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      && loc.getLongitude() <= this.getYMax());
+  }
+
+  /* 
+  * Checks if bounding box intersects with circle area of center location and 
+  * distance to closest point 
+  */
+  public boolean intersectsCircle(Location center, double distance) {
+    double boundaryX = (this.xMin + this.xMax)/2;
+    double boundaryY = (this.yMin + this.yMax)/2;
```
Comment (nicorod-google) [inline capstone/src/main/java/com/google/sps/Objects/BoundingBox.java]: nit: add a space before and after the division

File: capstone/src/main/java/com/google/sps/Objects/BoundingBox.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    if (circleDistanceX <= boundaryWidth/2) {
+      return true;
+    }
+    if (circleDistanceY <=boundaryHeight/2) {
+      return true;
+    }
+
+    double cornerDistance = (circleDistanceX - boundaryWidth/2)*(circleDistanceX - boundaryWidth/2) + 
+				(circleDistanceY - boundaryHeight/2)*(circleDistanceY - boundaryHeight/2);
+    return (cornerDistance <= distance*distance);
```
Comment (nicorod-google) [inline capstone/src/main/java/com/google/sps/Objects/BoundingBox.java]: nit: add spaces before and after the / and *

File: capstone/src/test/java/com/google/sps/objects/QuadTreeTest.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    assertTrue(quadTree.getChildren().size() == 0);
+    
+    assertTrue(quadTree.getLocations().contains(LOC_1));
+    assertTrue(quadTree.getLocations().contains(LOC_2));
+    assertTrue(quadTree.getLocations().contains(LOC_3));
+    assertTrue(quadTree.getLocations().contains(LOC_4));
+  }
+
+  @Test 
+  public void insertTest_oneSplit() {
```
Comment (nicorod-google) [inline capstone/src/test/java/com/google/sps/objects/QuadTreeTest.java:83]: Could you also add a test that has two splits? Mainly just to help me better understand how the quad tree is being constructed.

File: capstone/src/main/java/com/google/sps/Objects/QuadTree.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        numPoints ++;
+        return true;
+      }
+    }
+
+    // Node is not in this bounding box
+    return false;
+  }
+
+  void split() {
```
Comment (nicorod-google) [inline capstone/src/main/java/com/google/sps/Objects/QuadTree.java:90]: Can you add comments to these methods to describe what they are all doing? There is a lot of math happening so it would also be good to make sure it's documented well in your design doc

**Review by nicorod-google — CHANGES_REQUESTED** [review]

File: capstone/src/main/webapp/index.html (hunk 2 above)
Comment (anikabagga) [reply capstone/src/main/webapp/index.html]: Removing google maps stuff for this PR, will add to next PR

File: capstone/src/test/java/com/google/sps/objects/QuadTreeTest.java (hunk 7 above)
Comment (anikabagga) [reply capstone/src/test/java/com/google/sps/objects/QuadTreeTest.java:83]: added a test for two splits

File: capstone/src/main/java/com/google/sps/Objects/QuadTree.java (hunk 8 above)
Comment (anikabagga) [reply capstone/src/main/java/com/google/sps/Objects/QuadTree.java:90]: Added some comments, let me know if I can make anything more clear!

File: capstone/src/test/java/com/google/sps/objects/QuadTreeTest.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+              .setDefaultHighRepJobPolicyUnappliedJobPercentage(0));
+  
+  private static final Location LOC_1 = new Location("Test1", "address1", new Coordinate(40.48862915, -80.14141311));
+  private static final Location LOC_2 = new Location("Test2", "address2", new Coordinate(40.50170965, -80.09777729));
+  private static final Location LOC_3 = new Location("Test3", "address3", new Coordinate(40.57455069, -80.00992621)); 
+  private static final Location LOC_4 = new Location("Test4", "address4", new Coordinate(40.29651063, -79.98241256)); 
+  private static final Location LOC_5 = new Location("Test5", "address5", new Coordinate(40.49303469, -80.10254273)); 
+  private static final Location LOC_6 = new Location("Test6", "address6", new Coordinate(40.50245938, -80.08270203)); 
+  private static final Location LOC_7 = new Location("Test7", "address7", new Coordinate(40.4513275, -79.86228567)); 
+  private static final Location LOC_8 = new Location("Test8", "address8", new Coordinate(40.41427404,	-79.99527299));
```
Comment (g-asweet) [inline capstone/src/test/java/com/google/sps/objects/QuadTreeTest.java]: Personally, I would lean toward unrealistic, easy to follow values (40, -80), (40, -80.0000001), (40.0001, -80)

this will allow you to easily define the edge cases for distances where you can more confidently claim something is X far away, and hopefully end up with cleaner results that are easier to follow

Comment (g-asweet) [reply capstone/src/test/java/com/google/sps/objects/QuadTreeTest.java]: I'd also maybe consider instead of defining LOC's that you can't easily see IN the test themselves, a helper function:

```java
addTestLocation(x, y);
```

so you can easily see the coordinates you're comparing in the test and a developer at a glance should be able to tell something like (30, 30) is closer to (30, 40) than to (50, 30)

File: capstone/src/main/java/com/google/sps/Objects/QuadTree.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    return nearestLocations;
+  }
+
+  public PriorityQueue<Location> findKNearestNeighborsHelper(Location loc, int k, PriorityQueue<Location> pq) {
+
+    if (locations.size() == 0 || loc == null) {
+      return pq;
+    }
+
+    Location maxClosest = pq.peek();
```
Comment (g-asweet) [inline capstone/src/main/java/com/google/sps/Objects/QuadTree.java]: `maxClosest` is an odd name to me. Is this the `closestLocation`? `nearestLocation`?

File: capstone/src/main/java/com/google/sps/Objects/QuadTree.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    ArrayList<Location> nearestLocations = new ArrayList<>();
+
+    while (!pq.isEmpty()) {
+      nearestLocations.add(0, pq.poll());
+    }
+
+    return nearestLocations;
+  }
+
+  public PriorityQueue<Location> findKNearestNeighborsHelper(Location loc, int k, PriorityQueue<Location> pq) {
```
Comment (g-asweet) [inline capstone/src/main/java/com/google/sps/Objects/QuadTree.java]: instead of `pq` I'd consider `closestLocationPQ` or `nearestPQ` or `farthestPQ` something that tells me what is being prioritized

File: capstone/src/main/java/com/google/sps/Objects/QuadTree.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+    // Exceeded the capacity so split into four quadrants
+    if (topLeftTree == null) {
+      split();
+    }
+
+    // Add point to correct quadrant 
+    for (QuadTree qt: children) {
+      if (qt.insert(node)) {
+        numPoints ++;
```
Comment (g-asweet) [inline capstone/src/main/java/com/google/sps/Objects/QuadTree.java]: `numPoints++;`

**Review by g-asweet — DISMISSED** [review]

**Review by nicorod-google — COMMENTED** [review]
Ping me when you address Andrew's comments, I'll take another look and LGTM

**Review by nicorod-google — APPROVED** [review]
